File: solutions/0/0/5/9/p0059_s0.c
/* Solution to Euler Problem 59: XOR Decryption. */
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

char *get_text_file(const char *src) {
    const char *slash = strrchr(src, '/');
    const char *name_start = slash ? slash + 1 : src;
    const char *q = strchr(name_start, '?');
    size_t name_len = q ? (size_t)(q - name_start) : strlen(name_start);

    char exe_path[4096];
    ssize_t len = readlink("/proc/self/exe", exe_path, sizeof(exe_path) - 1);
    if (len < 0) return NULL;
    exe_path[len] = '\0';

    char path[4096];
    int pn = snprintf(path, sizeof(path), "%s/resources/%.*s", dirname(exe_path), (int)name_len, name_start);
    if (pn < 0 || (size_t)pn >= sizeof(path)) return NULL;

    FILE *f = fopen(path, "rb");
    if (!f) return NULL;
    if (fseek(f, 0, SEEK_END) != 0) { fclose(f); return NULL; }
    long sz = ftell(f);
    if (sz < 0) { fclose(f); return NULL; }
    rewind(f);
    char *buf = malloc((size_t)sz + 1);
    if (!buf) { fclose(f); return NULL; }
    if (fread(buf, 1, (size_t)sz, f) != (size_t)sz) { free(buf); fclose(f); return NULL; }
    buf[sz] = '\0';
    fclose(f);
    return buf;
}

long long solve(int argc, char *argv[]) {
    if (argc < 4) return -1;

    const char *file_url    = argv[1];
    int         key_length  = atoi(argv[2]);
    const char *most_common = argv[3];

    char *contents = get_text_file(file_url);
    if (!contents) return -1;

    /* Parse comma-separated integers */
    int capacity = 8192;
    int *encrypted = malloc((size_t)capacity * sizeof(int));
    if (!encrypted) { free(contents); return -1; }
    int n = 0;

    char *ptr = contents;
    while (*ptr) {
        while (*ptr == ' ' || *ptr == '\n' || *ptr == '\r' || *ptr == '\t') ptr++;
        if (*ptr == '\0') break;
        int val = 0, got = 0;
        while (*ptr >= '0' && *ptr <= '9') {
            val = val * 10 + (*ptr - '0');
            ptr++;
            got = 1;
        }
        if (got) {
            if (n >= capacity) {
                capacity *= 2;
                int *tmp = realloc(encrypted, (size_t)capacity * sizeof(int));
                if (!tmp) { free(encrypted); free(contents); return -1; }
                encrypted = tmp;
            }
            encrypted[n++] = val;
        }
        if (*ptr == ',') ptr++;
    }
    free(contents);

    if (n == 0 || key_length <= 0) { free(encrypted); return -1; }

    /* Build interleaved slices */
    int *slice_sizes = calloc((size_t)key_length, sizeof(int));
    if (!slice_sizes) { free(encrypted); return -1; }

    for (int j = 0; j < n; j++) slice_sizes[j % key_length]++;

    int **slices = malloc((size_t)key_length * sizeof(int *));
    if (!slices) { free(slice_sizes); free(encrypted); return -1; }

    for (int i = 0; i < key_length; i++) {
        slices[i] = malloc((size_t)(slice_sizes[i] ? slice_sizes[i] : 1) * sizeof(int));
        if (!slices[i]) {
            for (int k = 0; k < i; k++) free(slices[k]);
            free(slices); free(slice_sizes); free(encrypted); return -1;
        }
        slice_sizes[i] = 0;
    }
    for (int j = 0; j < n; j++) {
        int i = j % key_length;
        slices[i][slice_sizes[i]++] = encrypted[j];
    }
    free(encrypted);

    int common_len = (int)strlen(most_common);

    int *best_key   = calloc((size_t)key_length, sizeof(int));
    int *best_score = calloc((size_t)key_length, sizeof(int));
    if (!best_key || !best_score) {
        for (int i = 0; i < key_length; i++) free(slices[i]);
        free(slices); free(slice_sizes); free(best_key); free(best_score);
        return -1;
    }

    for (int kb = 97; kb <= 122; kb++) {
        for (int i = 0; i < key_length; i++) {
            int score = 0;
            int sz = slice_sizes[i];
            for (int j = 0; j < sz; j++) {
                char c = (char)(slices[i][j] ^ kb);
                for (int m = 0; m < common_len; m++) {
                    if (c == most_common[m]) { score++; break; }
                }
            }
            if (score > best_score[i]) {
                best_score[i] = score;
                best_key[i]   = kb;
            }
        }
    }

    /* Sum all ASCII values of decrypted text */
    long long total = 0;
    for (int i = 0; i < key_length; i++) {
        int kb = best_key[i];
        int sz = slice_sizes[i];
        for (int j = 0; j < sz; j++) {
            total += (slices[i][j] ^ kb);
        }
        free(slices[i]);
    }

    free(slices);
    free(slice_sizes);
    free(best_key);
    free(best_score);

    return total;
}

int main(int argc, char *argv[]) {
    int runs = 1;

    char **solve_argv = malloc((size_t)argc * sizeof(char *));
    if (!solve_argv) {
        fprintf(stderr, "runner: out of memory\n");
        return 1;
    }
    int solve_argc = 0;
    solve_argv[solve_argc++] = argv[0];

    for (int i = 1; i < argc; i++) {
        if (argv[i][0] == '\0') continue;
        if (strncmp(argv[i], "--runs=", 7) == 0) {
            int r = atoi(argv[i] + 7);
            if (r >= 1) runs = r;
            continue;
        }
        if (strcmp(argv[i], "--show") == 0) continue;
        solve_argv[solve_argc++] = argv[i];
    }

    long long result = 0;
    double total = 0.0;
    int rc = 0;
    int has_result = 0;

    for (int r = 0; r < runs; r++) {
        struct timespec t0, t1;
        clock_gettime(CLOCK_MONOTONIC, &t0);
        long long cur = solve(solve_argc, solve_argv);
        clock_gettime(CLOCK_MONOTONIC, &t1);
        total += (double)(t1.tv_sec - t0.tv_sec)
               + (double)(t1.tv_nsec - t0.tv_nsec) * 1e-9;
        if (has_result && cur != result) {
            fprintf(stderr, "Expected consistent result, got %lld previous result=%lld\n",
                    cur, result);
            rc = 1;
        }
        result = cur;
        has_result = 1;
    }

    free(solve_argv);
    printf("%d %.17g %lld\n", runs, total / (double)runs, result);
    return rc;
}

File: solutions/0/0/5/9/p0059_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 59: XOR Decryption [Level 1]. """
from __future__ import annotations

import sys
from pathlib import Path
from sys import argv, stderr
from time import perf_counter
from typing import Any


def get_text_file(src: str) -> str:
    """Return the contents of a file from the 'resources' directory."""
    local_filename: str = "resources/" + src.split("/")[-1].split("?")[0]
    return (Path(__file__).parent / local_filename).read_text()


def solve(*, file_url: str, key_length: int, most_common_letters: str) -> int:
    encrypted: list[int] = [int(x) for x in get_text_file(file_url).split(",")]
    slices_range: tuple[int, ...] = tuple(range(key_length))
    encrypted_slices: list[list[int]] = [encrypted[i::key_length] for i in slices_range]
    key: list[int] = [0] * key_length
    score: list[int] = [0] * key_length
    decrypted: list[str] = [""] * key_length
    for _key in range(97, 123):
        for i in slices_range:
            _decrypted = "".join((chr(_char ^ _key) for _char in encrypted_slices[i]))
            if (_score := sum((_decrypted.count(x) for x in most_common_letters))) > score[i]:
                score[i], key[i], decrypted[i] = (_score, _key, _decrypted)
    if sys.argv[-1] == "--show":
        print("".join(("".join(chars) for chars in zip(*decrypted))))
    return sum((ord(_char) for _decrypted in decrypted for _char in _decrypted))


def main(**kwargs: Any) -> int:
    """
    Usage: ./file.py <kwarg>... [--runs=1] [--show]
    Output: "<runs> <avg_seconds> <result>"
    """
    try:
        runs_arg: str = next((arg for arg in argv[1:] if arg.startswith("--runs=")))
        runs: int = int(runs_arg.split("=", 1)[1])
        assert runs > 0
    except (AssertionError, StopIteration, ValueError):
        runs = 1
    elapsed: list[float] = []
    result: int | None = None
    rc: int = 0
    errors: list[str] = []
    for _ in range(runs):
        _start, _result, _stop = (perf_counter(), solve(**kwargs), perf_counter())
        elapsed.append(_stop - _start)
        if result is not None and _result != result:
            errors.append(f"Expected consistent result, got {_result} previous result={result}")
        result = _result
    if result is None:
        errors.append("Expected a result, got None")
    average: float = sum(elapsed) / len(elapsed)
    if errors:
        print("\n".join(errors), file=stderr)
        rc = 1
    print(f"{runs} {average} {result}")
    return rc


if __name__ == "__main__":
    raise SystemExit(main(file_url=str(argv[1]), key_length=int(argv[2]), most_common_letters=str(argv[3])))
